styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -110,53 +124,76 @@ private void triggerBackfill(Backfill backfill, Map<String, Long> backfillStates
 
     final Workflow workflow = workflowOpt.get();
 
+    while (true) {
+      try {
+        if (!storage.runInTransaction(tx -> tryTriggerAndProgress(tx, backfill.id(), workflow))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("transaction failure when working on backfill {}", backfill, e);
+        break;
+      }
+    }
+  }
+
+  private Boolean tryTriggerAndProgress(StorageTransaction tx,
+                                        String id, Workflow workflow) {
+    final Backfill momentBackfill =
+        storage.backfill(id).orElseThrow(RuntimeException::new);
+
+    final Instant partition = momentBackfill.nextTrigger();
     final int remainingCapacity =
-        backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+        momentBackfill.concurrency() - stateManager.activeStates(id).size();
```

Comment:
I am a bit concerned about this best effort solution.
1 - Could be confusing for users to set limit 2 and see 3 instances being triggered
2 - Sometimes concurrency 1 is used to ensure strict sequential advancement of the backfill, which would now not be guaranteed anymore.

Reply:
I thought about it and it is actually not possible, otherwise it means two nodes can work on the same backfill concurrently and they both succeed without any conflict.

Reply:
I think it can also happen if the query from `activeStates` uses old indexes and doesn't see all the triggered backfill.

Reply:
That won’t matter I think. Even we calculate wron here, it won’t happen node A triggers partition 1 and node B triggers partition 2 concurrently. Or?

Reply:
From my tests on staging this seems to happen with even just one instance, i.e. one more WFI is triggered over the concurrency limit (it happens consistently). I assume that after the last expected trigger we call `tryTriggerAndProgress` one more time and we don't see the previously triggered WFI yet from `stateManager.activeStates(id)`

Reply:
It turned out datastore query is eventually consistent, so this solution doesn’t work. It’s probably better just keeping the counter in memory. Unless one node is super slow after computing the remaining capacity, race will be very rare.